Fix issues relating to features not matching graduated ranges

nyalldawson opened the pull request:
Caused by exact matching double values, when we need to have a little bit of tolerance. Also implements a fix in the OGR provider where certain methods (minimum/maximum/unique values) was force converting values via a string, resulting in loss of precision)

nyalldawson committed:
Cleanup duplicate code and add tests for graduated range resolving

nyalldawson committed:
Use real field type instead of a forced conversion through string
for OGR provider minimum/maximum/unique value retrieval

Avoids precision loss due to this string conversion, notably resulting
in graduated ranges which don't quite encompass the full range of
values present in a layer.

Fixes #32667, #27420

nyalldawson committed:
Add a bit of tolerance to graduated symbol range resolving

If a value doesn't fall exactly within a range, then try to see
if it falls just outside of any ranges (within double precision
tolerance), to correctly handle double precision values coming
from different sources.

Fixes #27420

nirvn commented:
Oh nice one!

nyalldawson commented:
@elpaso have you any idea why this causes a failure in the server? https://travis-ci.org/qgis/QGIS/jobs/607990439

elpaso commented:
@nyalldawson this patch breaks the temporal extent calculation, that is now `null` instead of the expected:

```
"temporal": {
      "interval": [
        [
          "2010-01-01T00:00:00",
          "2022-01-01T01:01:01Z"
        ]
      ],
      "trs": "http://www.opengis.net/def/uom/ISO-8601/0/Gregorian"
    }
```

the calculation is implemented in `QgsServerApiUtils::temporalExtent` but I cannot dig into it now and see what's going wrong with this patch.

nyalldawson committed:
[ogr] When calculating min/max value for date/datetime fields, we have
to use the older string conversion methods as OGR will always return
a string value for these aggregate queries, regardless of the original
data type

nyalldawson commented:
@rouault Does https://github.com/qgis/QGIS/pull/32677/commits/42a83c5891e1ea564a8674894ea4c3cb943a76d5 look correct to you? Seems that the OGR min/max aggregate results for date/datetime fields returns a string value, regardless of the original field data type. This issue is vaguely familiar and I have a feeling it was discussed sometime in the past, but can't find the relevant conversation now...

nyalldawson commented:
Anyone care to review this one?

nirvn reviewed (approved):
@nyalldawson , went through the PR, looking good. I've stumbled on this irritating issue before myself, glad to see it fixed.

nyalldawson commented:
Thanks @nirvn!
